framework/0000-spa-authentication.md

```diff
@@ -0,0 +1,142 @@
+  a CSRF token. It can optionally contain more data useful to the client-side app, e.g. username, email and roles.
+  Both can be achieved using a JWT token, to act as CSRF token but also carry a payload.
+
+Client-side application can then access to the new cookie but cannot access the session cookie, which is protected.
+On each request with a payload, the client-side app must send the CSRF token in the request (payload or header).
+
+Any other view provided by `Flask-Security` that authenticates user and performs a render_template or redirection
+should implemented in the same way: `logout`, `register`, `psw reset`, `change psw`, etc...
+
+#### Login via OAuth
+
+Invenio implements OAuth login with the module `invenio-oauthclient`. Create a new endpoint `GET /api/oauth-providers/`
+to expose all configured OAuth providers and allow client-side apps to render a login view with all available
+authentication methods.
+
+Create a new endpoint `POST /api/login/<remote_app>` to handle the OAuth flow: on request, return a redirection to the
+OAuth provider to perform the first part of the OAuth flow setting as callback parameter the backend authorized
+url `/api/authorized/`.
+On callback called, `invenio-oauthclient` should complete the OAuth flow as already implemented, but perform the same
+operations of double cookie creation as described before. It should then redirect back to the client-side app with a
+success/failure status.
+
+### CSRF
+
+Any HTTP request that sends a payload and relies on cookies authentication should be protected from CSRF attacks.
+The backend, for cookie authenticated requests only (not token based ones), should require and validate the CSRF token
+in the payload or as HTTP header.
+
+All Invenio REST endpoints should be reviewed to require CSRF token.
+Note: this is a **backward incompatible** change!
+
+### CSRF expiration
+
+For normal server side apps, the expiration of the CSRF token is normally quite short because it is injected with the
+form at rendering time, with the assumption that the form will be submitted shortly.
+For client-side apps, the expiration should be set longer, e.g. 30 minutes, because the backend cannot guess when a
+request will be submitted.
+
+To achieve so, the CSRF validity should be set with longer expiration and a new token should be generated on each
+POST/PUT request to ensure that a fresh one is always available to the client-side app.
+In case the CSRF token validation fails, the backend should provide a new token in the response.
+
+### SameSite cookie
+
+[SameSite cookie](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies) is a feature that allow servers to
+instrument browser to send cookies only to requests to the same site, and not to send them cross-site.
+This will enable protection against CSRF.
+Unfortunately, the `SameSite` feature is relatively new and it is supported only by recent browsers: it cannot
```

Comment:
I would also add a link to [CanIUse for the feature](https://caniuse.com/#feat=same-site-cookie-attribute)